Support new clip ufunc

Starting with numpy 1.17, `np.clip` will be based on a `ufunc`, which means we can ensure it works properly with `Quantity`. (Until we do so, we might also get `numpy-dev` failures.)

@mhvk , numpy-dev is failing now; e.g. https://travis-ci.org/astropy/astropy/jobs/536308798

```
________________________ TestUfuncHelpers.test_coverage ________________________
self = <astropy.units.tests.test_quantity_ufuncs.TestUfuncHelpers object at 0x7f11069a17b8>
    def test_coverage(self):
        """Test that we cover all ufunc's"""
    
        all_np_ufuncs = set([ufunc for ufunc in np.core.umath.__dict__.values()
                             if isinstance(ufunc, np.ufunc)])
    
        all_q_ufuncs = (qh.UNSUPPORTED_UFUNCS |
                        set(qh.UFUNC_HELPERS.keys()))
        # Check that every numpy ufunc is covered.
>       assert all_np_ufuncs - all_q_ufuncs == set()
E       AssertionError: assert {<ufunc 'clip'>} == set()
E         Extra items in the left set:
E         <ufunc 'clip'>
E         Use -v to get the full diff
astropy/units/tests/test_quantity_ufuncs.py:69: AssertionError
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -311,6 +311,10 @@ astropy.uncertainty
 astropy.units
 ^^^^^^^^^^^^^
 
+- Add support for the ``clip`` ufunc, which in numpy 1.17 is used to implement
+  ``np.clip``.  As part of that, remove the ``Quantity.clip`` method under
+  numpy 1.17. [#8747]
+
 astropy.utils
 ^^^^^^^^^^^^^
 

--- a/astropy/units/function/core.py
+++ b/astropy/units/function/core.py
@@ -683,3 +683,7 @@ class FunctionQuantity(Quantity):
 
     def cumsum(self, axis=None, dtype=None, out=None):
         return self._wrap_function(np.cumsum, axis, dtype, out=out)
+
+    def clip(self, a_min, a_max, out=None):
+        return self._wrap_function(np.clip, self._to_own_unit(a_min),
+                                   self._to_own_unit(a_max), out=out)

--- a/astropy/units/quantity.py
+++ b/astropy/units/quantity.py
@@ -20,7 +20,7 @@ from .core import (Unit, dimensionless_unscaled, get_current_unit_registry,
                    UnitBase, UnitsError, UnitConversionError, UnitTypeError)
 from .utils import is_effectively_unity
 from .format.latex import Latex
-from astropy.utils.compat import NUMPY_LT_1_14, NUMPY_LT_1_16
+from astropy.utils.compat import NUMPY_LT_1_14, NUMPY_LT_1_16, NUMPY_LT_1_17
 from astropy.utils.compat.misc import override__dir__
 from astropy.utils.exceptions import AstropyDeprecationWarning, AstropyWarning
 from astropy.utils.misc import isiterable, InheritDocstrings
@@ -455,9 +455,10 @@ class Quantity(np.ndarray, metaclass=InheritDocstrings):
             kwargs['out'] = (out_array,) if function.nout == 1 else out_array
 
         # Same for inputs, but here also convert if necessary.
-        arrays = [(converter(input_.value) if converter else
-                   getattr(input_, 'value', input_))
-                  for input_, converter in zip(inputs, converters)]
+        arrays = []
+        for input_, converter in zip(inputs, converters):
+            input_ = getattr(input_, 'value', input_)
+            arrays.append(converter(input_) if converter else input_)
 
         # Call our superclass's __array_ufunc__
         result = super().__array_ufunc__(function, method, *arrays, **kwargs)
@@ -1502,9 +1503,10 @@ class Quantity(np.ndarray, metaclass=InheritDocstrings):
         result = function(*args, **kwargs)
         return self._result_as_quantity(result, unit, out)
 
-    def clip(self, a_min, a_max, out=None):
-        return self._wrap_function(np.clip, self._to_own_unit(a_min),
-                                   self._to_own_unit(a_max), out=out)
+    if NUMPY_LT_1_17:
+        def clip(self, a_min, a_max, out=None):
+            return self._wrap_function(np.clip, self._to_own_unit(a_min),
+                                       self._to_own_unit(a_max), out=out)
 
     def trace(self, offset=0, axis1=0, axis2=1, dtype=None, out=None):
         return self._wrap_function(np.trace, offset, axis1, axis2, dtype,

--- a/astropy/units/quantity_helper/converters.py
+++ b/astropy/units/quantity_helper/converters.py
@@ -166,31 +166,34 @@ def converters_and_unit(function, method, *args):
         converters, result_unit = ufunc_helper(function, *units)
 
         if any(converter is False for converter in converters):
-            # for two-argument ufuncs with a quantity and a non-quantity,
+            # for multi-argument ufuncs with a quantity and a non-quantity,
             # the quantity normally needs to be dimensionless, *except*
             # if the non-quantity can have arbitrary unit, i.e., when it
             # is all zero, infinity or NaN.  In that case, the non-quantity
             # can just have the unit of the quantity
             # (this allows, e.g., `q > 0.` independent of unit)
-            maybe_arbitrary_arg = args[converters.index(False)]
             try:
-                if can_have_arbitrary_unit(maybe_arbitrary_arg):
-                    converters = [None, None]
-                else:
-                    raise UnitConversionError(
-                        "Can only apply '{0}' function to "
-                        "dimensionless quantities when other "
-                        "argument is not a quantity (unless the "
-                        "latter is all zero/infinity/nan)"
-                        .format(function.__name__))
+                # Don't fold this loop in the test above: this rare case
+                # should not make the common case slower.
+                for i, converter in enumerate(converters):
+                    if converter is not False:
+                        continue
+                    if can_have_arbitrary_unit(args[i]):
+                        converters[i] = None
+                    else:
+                        raise UnitConversionError(
+                            "Can only apply '{0}' function to "
+                            "dimensionless quantities when other "
+                            "argument is not a quantity (unless the "
+                            "latter is all zero/infinity/nan)"
+                            .format(function.__name__))
             except TypeError:
                 # _can_have_arbitrary_unit failed: arg could not be compared
                 # with zero or checked to be finite. Then, ufunc will fail too.
                 raise TypeError("Unsupported operand type(s) for ufunc {0}: "
-                                "'{1}' and '{2}'"
-                                .format(function.__name__,
-                                        args[0].__class__.__name__,
-                                        args[1].__class__.__name__))
+                                "'{1}'".format(function.__name__,
+                                               ','.join([arg.__class__.__name__
+                                                         for arg in args])))
 
         # In the case of np.power and np.float_power, the unit itself needs to
         # be modified by an amount that depends on one of the input values,

--- a/astropy/units/quantity_helper/helpers.py
+++ b/astropy/units/quantity_helper/helpers.py
@@ -296,6 +296,39 @@ def helper_divmod(f, unit1, unit2):
     return converters, (dimensionless_unscaled, result_unit)
 
 
+def helper_clip(f, unit1, unit2, unit3):
+    # Treat the array being clipped as primary.
+    converters = [None]
+    if unit1 is None:
+        result_unit = dimensionless_unscaled
+        try:
+            converters += [(None if unit is None else
+                            get_converter(unit, dimensionless_unscaled))
+                           for unit in (unit2, unit3)]
+        except UnitsError:
+            raise UnitConversionError(
+                "Can only apply '{0}' function to quantities with "
+                "compatible dimensions".format(f.__name__))
+
+    else:
+        result_unit = unit1
+        for unit in unit2, unit3:
+            try:
+                converter = get_converter(_d(unit), result_unit)
+            except UnitsError:
+                if unit is None:
+                    # special case: OK if unitless number is zero, inf, nan
+                    converters.append(False)
+                else:
+                    raise UnitConversionError(
+                        "Can only apply '{0}' function to quantities with "
+                        "compatible dimensions".format(f.__name__))
+            else:
+                converters.append(converter)
+
+    return converters, result_unit
+
+
 # list of ufuncs:
 # http://docs.scipy.org/doc/numpy/reference/ufuncs.html#available-ufuncs
 
@@ -407,3 +440,6 @@ UFUNC_HELPERS[np.floor_divide] = helper_twoarg_floor_divide
 UFUNC_HELPERS[np.heaviside] = helper_heaviside
 UFUNC_HELPERS[np.float_power] = helper_power
 UFUNC_HELPERS[np.divmod] = helper_divmod
+# Check for clip ufunc; note that np.clip is a wrapper function, not the ufunc.
+if isinstance(getattr(np.core.umath, 'clip', None), np.ufunc):
+    UFUNC_HELPERS[np.core.umath.clip] = helper_clip

--- a/astropy/units/tests/test_quantity.py
+++ b/astropy/units/tests/test_quantity.py
@@ -496,11 +496,10 @@ class TestQuantityOperations:
 
     def test_non_number_type(self):
         q1 = u.Quantity(11.412, unit=u.meter)
-        type_err_msg = ("Unsupported operand type(s) for ufunc add: "
-                        "'Quantity' and 'dict'")
         with pytest.raises(TypeError) as exc:
             q1 + {'a': 1}
-        assert exc.value.args[0] == type_err_msg
+        assert exc.value.args[0].startswith(
+            "Unsupported operand type(s) for ufunc add:")
 
         with pytest.raises(TypeError):
             q1 + u.meter

--- a/astropy/units/tests/test_quantity_ufuncs.py
+++ b/astropy/units/tests/test_quantity_ufuncs.py
@@ -868,6 +868,92 @@ class TestInplaceUfuncs:
             a4 += u.Quantity(10, u.mm, dtype=np.int64)
 
 
+@pytest.mark.skipif(not hasattr(np.core.umath, 'clip'),
+                    reason='no clip ufunc available')
+class TestClip:
+    """Test the clip ufunc.
+
+    In numpy, this is hidden behind a function that does not backwards
+    compatibility checks.  We explicitly test the ufunc here.
+    """
+    def setup(self):
+        self.clip = np.core.umath.clip
+
+    def test_clip_simple(self):
+        q = np.arange(-1., 10.) * u.m
+        q_min = 125 * u.cm
+        q_max = 0.0055 * u.km
+        result = self.clip(q, q_min, q_max)
+        assert result.unit == q.unit
+        expected = self.clip(q.value, q_min.to_value(q.unit),
+                             q_max.to_value(q.unit)) * q.unit
+        assert np.all(result == expected)
+
+    def test_clip_unitless_parts(self):
+        q = np.arange(-1., 10.) * u.m
+        qlim = 0.0055 * u.km
+        # one-sided
+        result1 = self.clip(q, -np.inf, qlim)
+        expected1 = self.clip(q.value, -np.inf, qlim.to_value(q.unit)) * q.unit
+        assert np.all(result1 == expected1)
+        result2 = self.clip(q, qlim, np.inf)
+        expected2 = self.clip(q.value, qlim.to_value(q.unit), np.inf) * q.unit
+        assert np.all(result2 == expected2)
+        # Zero
+        result3 = self.clip(q, np.zeros(q.shape), qlim)
+        expected3 = self.clip(q.value, 0, qlim.to_value(q.unit)) * q.unit
+        assert np.all(result3 == expected3)
+        # Two unitless parts, array-shaped.
+        result4 = self.clip(q, np.zeros(q.shape), np.full(q.shape, np.inf))
+        expected4 = self.clip(q.value, 0, np.inf) * q.unit
+        assert np.all(result4 == expected4)
+
+    def test_clip_dimensionless(self):
+        q = np.arange(-1., 10.) * u.dimensionless_unscaled
+        result = self.clip(q, 200 * u.percent, 5.)
+        expected = self.clip(q, 2., 5.)
+        assert result.unit == u.dimensionless_unscaled
+        assert np.all(result == expected)
+
+    def test_clip_ndarray(self):
+        a = np.arange(-1., 10.)
+        result = self.clip(a, 200 * u.percent, 5. * u.dimensionless_unscaled)
+        assert isinstance(result, u.Quantity)
+        expected = self.clip(a, 2., 5.) * u.dimensionless_unscaled
+        assert np.all(result == expected)
+
+    def test_clip_quantity_inplace(self):
+        q = np.arange(-1., 10.) * u.m
+        q_min = 125 * u.cm
+        q_max = 0.0055 * u.km
+        expected = self.clip(q.value, q_min.to_value(q.unit),
+                             q_max.to_value(q.unit)) * q.unit
+        result = self.clip(q, q_min, q_max, out=q)
+        assert result is q
+        assert np.all(result == expected)
+
+    def test_clip_ndarray_dimensionless_output(self):
+        a = np.arange(-1., 10.)
+        q = np.zeros_like(a) * u.m
+        expected = self.clip(a, 2., 5.) * u.dimensionless_unscaled
+        result = self.clip(a, 200 * u.percent, 5. * u.dimensionless_unscaled,
+                           out=q)
+        assert result is q
+        assert result.unit == u.dimensionless_unscaled
+        assert np.all(result == expected)
+
+    def test_clip_errors(self):
+        q = np.arange(-1., 10.) * u.m
+        with pytest.raises(u.UnitsError):
+            self.clip(q, 0, 1*u.s)
+        with pytest.raises(u.UnitsError):
+            self.clip(q.value, 0, 1*u.s)
+        with pytest.raises(u.UnitsError):
+            self.clip(q, -1, 0.)
+        with pytest.raises(u.UnitsError):
+            self.clip(q, 0., 1.)
+
+
 class TestUfuncAt:
     """Test that 'at' method for ufuncs (calculates in-place at given indices)
 

--- a/astropy/utils/compat/numpycompat.py
+++ b/astropy/utils/compat/numpycompat.py
@@ -7,7 +7,7 @@ from astropy.utils import minversion
 
 
 __all__ = ['NUMPY_LT_1_14', 'NUMPY_LT_1_14_1', 'NUMPY_LT_1_14_2',
-           'NUMPY_LT_1_16']
+           'NUMPY_LT_1_16', 'NUMPY_LT_1_17']
 
 # TODO: It might also be nice to have aliases to these named for specific
 # features/bugs we're checking for (ex:
@@ -16,3 +16,4 @@ NUMPY_LT_1_14 = not minversion('numpy', '1.14')
 NUMPY_LT_1_14_1 = not minversion('numpy', '1.14.1')
 NUMPY_LT_1_14_2 = not minversion('numpy', '1.14.2')
 NUMPY_LT_1_16 = not minversion('numpy', '1.16')
+NUMPY_LT_1_17 = not minversion('numpy', '1.17')

